lib/types/application/ApplicationFormatter.js

```diff
@@ -19,12 +26,33 @@ class ApplicationFormatter extends AbstractFormatter {
 				"/": project.resources.configuration.paths.webapp
 			};
 
-			return this.readManifest(project).then(function(manifest) {
-				if (!manifest["sap.app"] || !manifest["sap.app"].id) {
+			return this.readManifest(project).then((manifest) => {
+				// check for a proper sap.app/id in manifest.json to determine namespace
+				const appId = manifest["sap.app"] && manifest["sap.app"].id;
+				if (!appId) {
 					log.warn(`No "sap.app" ID configuration found in manifest of project ${project.metadata.name}`);
 					return;
 				}
-				project.metadata.namespace = manifest["sap.app"].id.replace(/\./g, "/");
+				// check app id for placeholder and try to read the placeholder from the pom.xml
+				const parts = appId.match(/^\$\{(.*)\}$/);
+				if (parts) {
+					log.warn(`No valid "sap.app" ID found in manifest of project ${project.metadata.name}. Reading "${parts[1]}" from pom.xml...`);
+					return this.readPOM(project).then((pom) => {
+						if (pom.project && pom.project.properties && pom.project.properties[parts[1]]) {
+							return pom.project.properties[parts[1]];
+						} else {
+							let obj = pom;
+							parts[1].split(".").forEach((part) => {
+								obj = obj[part];
+								return !obj;
```

Comment:
I think this return is not necessary. According to [mdn](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach):

> There is no way to stop or break a forEach() loop other than by throwing an exception. If you need such behavior, the forEach() method is the wrong tool.

Reply:
I did bookmark this in my brain - now it is! Next time you will not see the return - might have been done wrong in several other places... :-)